include/flash.h

```diff
@@ -28,18 +29,57 @@
 extern "C" {
 #endif
 
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+struct flash_pg_layout {
+	u32_t pages_count; /* count of pages sequence of the same size */
+	size_t pages_size;
+};
+
+struct flash_page_info {
+	off_t start_offset;
+	size_t size;
+};
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
 typedef int (*flash_api_read)(struct device *dev, off_t offset, void *data,
 			      size_t len);
 typedef int (*flash_api_write)(struct device *dev, off_t offset,
 			       const void *data, size_t len);
 typedef int (*flash_api_erase)(struct device *dev, off_t offset, size_t size);
 typedef int (*flash_api_write_protection)(struct device *dev, bool enable);
 
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+/**
+ *  @brief  Get the layout of pages of flash memory.
+ *
+ *  Layout is provided in run-length format. For a flash memory which consists
+ *  of uniform pages, the layout will be simply the one couple of pages count
+ *  and their size.
+ *  For a flash memory which conststs of pages of different sizes, layout will
+ *  be the sequence of such couples.
+ *
+ *  @param  dev     :flash device
+ *  @param  group   :start of a couple index to read
+ *  @param  num     :number of couples to read
+ *  @param  layout  :array of couple structures for storing the layout
+ *
+ *  @return  Number of groups retrived. Could be less than the requested number.
+ */
+
+typedef u32_t (*flash_api_pages_layout)(struct device *dev,
+				      u32_t group,
```

Comment:
I think your Idea is ideal for that. I thought about similar API before - but had doubt regard whether "layout" would be implemented always as const array.